source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,353 @@
+    }
+
+    /***************************************************************************
+
+        Make an enrollment data for enrollment process
+
+        Params:
+            frozen_utxo_hash = the hash of a frozen UTXO used to identify 
+                                a validator and to generate a siging key
+
+        Returns:
+            an enrollment data
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (const ref Hash frozen_utxo_hash) @trusted
+    {
+        static ubyte[] buffer;
+        buffer.length = 0;
+
+        // K, frozen UTXO hash
+        this.data.utxo_key = frozen_utxo_hash;
+
+        // N, cycle length
+        this.data.cycle_length = 1008; // freezing period / 2
+
+        // generate random seed value
+        this.random_seed_src = Scalar.random();
+
+        // X, nth image of random seed
+        this.data.random_seed = hashFull(this.random_seed_src);
+        foreach (i; 0 .. this.data.cycle_length-1)
+            this.data.random_seed = hashFull(this.data.random_seed);
+
+        // R, signature noise
+        this.signature_noise = Pair.random();
+        () @trusted { assumeSafeAppend(buffer); }();
+
+        scope SerializeDg dg = (scope const(ubyte[]) data) nothrow @safe
+        {
+            buffer ~= data;
+        };
+
+        serializePart(this.signature_noise, dg);
+        
+        () @trusted {
+            auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM node_enroll_data " ~
+                "WHERE key = ?)", "signature_noise");
+            if (results.oneValue!(bool))
+            {
+                this.db.execute("UPDATE node_enroll_data SET val = ? WHERE key = ?",
+                    buffer, "signature_noise");
+            }
+            else
+            {
+                this.db.execute("INSERT INTO node_enroll_data (key, val) VALUES (?, ?)",
+                    "signature_noise", buffer);
+            }
+        }();
+
+        // signature
+        auto pk = Scalar(this.node_key_pair.secret);
+        auto pk_point = pk.toPoint();
+        data.enroll_sig = sign(pk, pk_point, this.signature_noise.V,
+            this.signature_noise.v, this.data);
+
+        return this.data;
+    }
+
+    /***************************************************************************
+
+        Check if the enrollment data exists in the validator set.
+
+        Params:
+            enroll_hash = key for an enrollment data which is hash of frozen UTXO 
+
+        Returns:
+            true if the enrollment manager has the enrollment data
+
+    ***************************************************************************/
+
+    public bool hasEnrollment (const ref Hash enroll_hash) @trusted
+    {
+        auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM validator_set " ~
+            "WHERE key = ?)", enroll_hash[]);
+
+        return results.oneValue!(bool);
```

Comment:
No need for the intermediate variable, return it directly (it might also be more optimal thanks to NRVO)